packages/fxa-settings/src/lib/context/implementations/storage-context.ts

```diff
@@ -0,0 +1,91 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { Context } from "..";
+
+export class StorageContext implements Context {
+
+  private static readonly NAMESPACE = '__fxa_session';
+  private static readonly PERSIST_TO_LOCAL_STORAGE = ['oauth'];
+
+  private state: Record<string, any>;
+
+  constructor(
+    private window: Window
+  ) {
+    this.state = {};
+    this.load();
+  }
+
+  public requiresSync() {
+    return true;
+  }
+
+  public load() {
+    let values = {};
+
+    // Try parsing sessionStorage values
+    try {
+      const sessionStorageValueRaw = this.window.sessionStorage.getItem(StorageContext.NAMESPACE);
+      if (sessionStorageValueRaw != null) {
+        values = {
+          ...values,
+          ...JSON.parse(sessionStorageValueRaw)
+        };
+      }
+    } catch (e) {
+      // ignore the parse error.
+    }
+
+    // Try parsing localStorage values
+    try {
+      const localStorageValueRaw = this.window.localStorage.getItem(StorageContext.NAMESPACE);
+      if (localStorageValueRaw != null) {
+        values = {
+          ...values,
+          ...JSON.parse(localStorageValueRaw)
+        };
+      }
+    } catch (e) {
+      // ignore the parse error.
+    }
+
+    this.state = values;
+  }
+
+  public persist() {
+
+    const toSaveToSessionStorage: Record<string, any> = {};
+    const toSaveToLocalStorage: Record<string, any> = {};
+
+    for (const key in this.state) {
+      const value = this.state[key];
+
+      if (StorageContext.PERSIST_TO_LOCAL_STORAGE.indexOf(key) >= 0) {
+        toSaveToLocalStorage[key] = value;
+      } else {
+        toSaveToSessionStorage[key] = value;
+      }
+    }
+
+    // Wrap browser storage access in a try/catch block because some browsers
+    // (Firefox, Chrome) except when trying to access browser storage and
+    // cookies are disabled.
+    try {
+      this.window.localStorage.setItem(StorageContext.NAMESPACE, JSON.stringify(toSaveToLocalStorage));
+      this.window.sessionStorage.setItem(StorageContext.NAMESPACE, JSON.stringify(toSaveToSessionStorage));
+    } catch (e) {
+      // some browsers disable access to browser storage
+      // if cookies are disabled.
+    }
+  }
+
+  public get(key: string) {
+    return this.state[key];
+  }
+
+  public set(key: string, value: any) {
```

Comment:
I did experimented with this a little bit. I'll revisit it again though. As I recall, I ran into a problem. At one point I actually had a generic in place (ie `set<T>(key:string, value:T)`), because I thought I could propagate the type all the way into this storage layer. This didn't end up working, but I don't recall why anymore.

It's good to note that because we use these storage contexts as a back end store for the `@bind` decorator, an interesting thing happens. The `@bind` decorator is inherently typed, and it also has validation that should enforce its type. 

Therefore we actually retain type safety as long as we only access the storage context through fields/properties that have been decorated with the `@bind` directive. Obviously if we go behind the bind decorator, all bets are off. 

Anyway, since the @bind decorator is type safe, I feel a bit better about the `any` type here. This store is essentially just one small abstraction layer between the model and the storage api (or other kv store). The storage API can't strong types anyways, so the fact we have a general `any` here doesn't bother me too much.